File: Day310_2025/Q310_Leetcode.cpp
class Vertex{
public:
    int vertexId;
    bool offline = false;
    int powerGridId = -1;

    Vertex(){}
    Vertex(int id) : vertexId(id){}
};

using PowerGrid = priority_queue<int, vector<int>, greater<int>>;
using Graph = vector<vector<int>>;

class Solution {
    vector<Vertex> vertices = vector<Vertex>();

    void traverse(Vertex& u, int powerGridId, PowerGrid& powerGrid, Graph& graph){
        u.powerGridId = powerGridId;
        powerGrid.push(u.vertexId);
        for(int vId : graph[u.vertexId]){
            Vertex& v = vertices[vId];
            if(v.powerGridId == -1){
                traverse(v, powerGridId, powerGrid, graph);
            }
        }
    }
public:
    vector<int> processQueries(int c, vector<vector<int>>& connections, vector<vector<int>>& queries) {
        // Q. Power Grid Maintenance
        Graph graph(c + 1);
        vertices.resize(c + 1);

        for(int index = 1; index <= c; ++index){
            vertices[index] = Vertex(index);
        }

        for(auto& connection : connections){
            graph[connection.at(0)].push_back(connection.at(1));
            graph[connection.at(1)].push_back(connection.at(0));
        }

        vector<PowerGrid> powerGrids;
        for(int index = 1, powerGridId = 0; index <= c; ++index){
            auto& v = vertices[index];
            if(v.powerGridId == -1){
                PowerGrid powerGrid;
                traverse(v, powerGridId, powerGrid, graph);
                powerGrids.push_back(powerGrid);
                powerGridId++;
            }
        }

        vector<int> answer;
        for(auto& query : queries){
            int operation = query.at(0), vertex = query.at(1);
            if(operation == 1){
                if(!vertices[vertex].offline){
                    answer.push_back(vertex);
                }
                else{
                    auto& powerGrid = powerGrids[vertices[vertex].powerGridId];
                    while(!powerGrid.empty() && vertices[powerGrid.top()].offline){
                        powerGrid.pop();
                    }
                    answer.push_back(!powerGrid.empty() ? powerGrid.top() : -1);
                }
            }
            else if(operation == 2){
                vertices[vertex].offline = true;
            }
        }

        return answer;
    }
};

File: Day310_2025/Q310_Leetcode.py
import heapq
from typing import List

class Vertex:
    def __init__(self, vertexId: int):
        self.vertexId = vertexId
        self.offline = False
        self.powerGridId = -1

class Solution:
    def processQueries(self, c: int, connections: List[List[int]], queries: List[List[int]]) -> List[int]:
        graph = [[] for _ in range(c + 1)]
        vertices = [Vertex(i) for i in range(c + 1)]

        for u, v in connections:
            graph[u].append(v)
            graph[v].append(u)

        def traverse(u: Vertex, grid_id: int, grid_heap: List[int]):
            u.powerGridId = grid_id
            heapq.heappush(grid_heap, u.vertexId)
            for vId in graph[u.vertexId]:
                v = vertices[vId]
                if v.powerGridId == -1:
                    traverse(v, grid_id, grid_heap)

        power_grids = []
        grid_id = 0
        for i in range(1, c + 1):
            if vertices[i].powerGridId == -1:
                grid_heap = []
                traverse(vertices[i], grid_id, grid_heap)
                power_grids.append(grid_heap)
                grid_id += 1

        answer = []
        for op, vertex in queries:
            v = vertices[vertex]
            if op == 1:
                if not v.offline:
                    answer.append(vertex)
                else:
                    pq = power_grids[v.powerGridId]
                    while pq and vertices[pq[0]].offline:
                        heapq.heappop(pq)
                    answer.append(pq[0] if pq else -1)
            elif op == 2:
                v.offline = True

        return answer
